### PR TITLE
ISPN-12036 Automatically register SerializationContextInitializers de…

### PR DESCRIPTION
**Pull request by tristantarrant**
https://issues.redhat.com/browse/ISPN-12036
Goes hand in hand with https://github.com/infinispan/protostream/pull/101

### CODE REVIEW

**Comment by anistor** [comment]
I remember at least another 2 other jira issues about the same topic. It would be nice to dedup them and have some description mentioning if this is only about client, server or both. 

And last, but not least we need to have a mechanism to not always load all providers from CP, but just a configured list (as  we already do). Is that supported by current approach ?

**Comment by ryanemerson** [comment]
> I remember at least another 2 other jira issues about the same topic. It would be nice to dedup them and have some description mentioning if this is only about client, server or both.

https://issues.redhat.com/browse/ISPN-11993 is the Jira to track changes for the client. It would be good to add more info to the ISPN-12036 Jira and commit indicating the service loading only affects the server.



> And last, but not least we need to have a mechanism to not always load all providers from CP, but just a configured list (as we already do). Is that supported by current approach ?

I believe that's why @tristantarrant has set `service = false` on our internal instances.

**Comment by tristantarrant** [comment]
Yes, the PR I created for Protostream sets `service=true' to default, as that is what I'd like for users, whereas all our internal SCIs have `false` which means it's up to us to register them manually as we (supposedly) know what we are doing

**Comment by ryanemerson** [comment]
Thanks @tristantarrant

**Comment by anistor** [comment]
Sorry for not providing more feedback on this, but I do not fully agree with the approach. Autoregistering all SCIs found in class path is a bad idea.

File: remote-query/remote-query-server/src/main/java/org/infinispan/query/remote/impl/ProtobufMetadataManagerImpl.java [hunk 1]
```diff
@@ -95,6 +95,7 @@ void start() {
             getProtobufMetadataCacheConfig(globalConfiguration).build(),
             EnumSet.of(InternalCacheRegistry.Flag.USER, InternalCacheRegistry.Flag.PROTECTED, InternalCacheRegistry.Flag.PERSISTENT));
 
+      processSerializationContextInitializer(ServiceFinder.load(SerializationContextInitializer.class, globalConfiguration.classLoader()));
```
Comment (anistor) [inline remote-query/remote-query-server/src/main/java/org/infinispan/query/remote/impl/ProtobufMetadataManagerImpl.java:98]: Providing more detail on what I said before, here I was expecting the service loader would be used only if there are no manually configured SCI's in ```globalConfiguration.serialization()```.

The auto-registration is nice for simple use cases, which might account for 90% real use cases. But it cannot give you any way out of situations when schemas are conflicting for any reason.

To support debugging of such situations I propose adding INFO logging here, with the names of all SCI's being registred. And after the registration I would also dump the eventual errors, with ERROR level immediately.

**Pull request merged**